### ISSUE TITLE
[BUG] Pip: not found when using mlflow build-docker locally

### ISSUE THREAD

**Issue by murilommen** [mlflow/mlflow#4561]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Mac OS Big Sur 11.3.1 (Apple M1 chip)
- **MLflow installed from (source or binary)**: PiPy
- **MLflow version (run ``mlflow --version``)**: `1.18.0`
- **Python version**: 3.8.5
- **npm version, if running the dev UI**:
- **Exact command to reproduce**: `mlflow models build-docker -m 'models:/{model_name}/{model_version}' -n my-mlflow-image`

### Describe the problem
I have tried to create a Docker image from a registered model in Mlflow (I am using the Databricks managed version here). The model was registered as a PyFunc.
It will download the model, identify its flavour and run until the `6/12` step, where it breaks with the following message:
```
#9 0.665 /bin/sh: 1: pip: not found
#9 ERROR: executor failed running [/bin/sh -c pip install mlflow==1.19.0]: exit code: 127
------
 > [ 6/12] RUN pip install mlflow==1.19.0:
------
executor failed running [/bin/sh -c pip install mlflow==1.19.0]: exit code: 127
```

### Code to reproduce issue
You must have a registered model or run in an Mlflow artifact storage. Then:
`mlflow models build-docker -m 'models:/{model_name}/{model_version}' -n my-mlflow-image` or `mlflow models build-docker -m 'runs:/{run_id}/model' -n my-mlflow-image` 

### Other info / logs
The full output from the command is as follows:

```bash
mlflow models build-docker -m 'runs:/{my_run}/model' -n direct-mlflow-image
2021/07/14 17:25:44 INFO mlflow.models.cli: Selected backend for flavor 'python_function'
2021/07/14 17:25:49 INFO mlflow.models.docker_utils: Building docker image with name direct-mlflow-image
/var/folders/yx/6fcq_tx96n52vr0kc66ypy9w0000gp/T/tmpyesejo4m/
/var/folders/yx/6fcq_tx96n52vr0kc66ypy9w0000gp/T/tmpyesejo4m//model_dir
/var/folders/yx/6fcq_tx96n52vr0kc66ypy9w0000gp/T/tmpyesejo4m//model_dir/model
/var/folders/yx/6fcq_tx96n52vr0kc66ypy9w0000gp/T/tmpyesejo4m//model_dir/model/MLmodel
/var/folders/yx/6fcq_tx96n52vr0kc66ypy9w0000gp/T/tmpyesejo4m//model_dir/model/model.pkl
/var/folders/yx/6fcq_tx96n52vr0kc66ypy9w0000gp/T/tmpyesejo4m//model_dir/model/input_example.json
/var/folders/yx/6fcq_tx96n52vr0kc66ypy9w0000gp/T/tmpyesejo4m//model_dir/model/conda.yaml
/var/folders/yx/6fcq_tx96n52vr0kc66ypy9w0000gp/T/tmpyesejo4m//Dockerfile
#1 [internal] load build definition from Dockerfile
#1 sha256:c4ce501e24{}
#1 transferring dockerfile: 1.58kB done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:7094b9fc79b9{}
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/ubuntu:18.04
#3 sha256:fbbba859d4013c87{}
#3 DONE 1.8s

#4 [ 1/12] FROM docker.io/library/ubuntu:18.04@sha256:3b8692dc4{}
#4 sha256:f601b492ef7f7c5a96{}
#4 DONE 0.0s

#5 [ 2/12] RUN apt-get -y update && apt-get install -y --no-install-recommends          wget          curl          nginx          ca-certificates          bzip2          build-essential          cmake          openjdk-8-jdk          git-core          maven     && rm -rf /var/lib/apt/lists/*
#5 sha256:28ae033a3a{}
#5 CACHED

#6 [ 3/12] RUN curl -L https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh >> miniconda.sh
#6 sha256:3dd39dfce1f4b9{}
#6 CACHED

#7 [ 4/12] RUN bash ./miniconda.sh -b -p /miniconda; rm ./miniconda.sh;
#7 sha256:15001db7c7586{}
#7 CACHED

#8 [ 5/12] WORKDIR /opt/mlflow
#8 sha256:7464451f9d65e5{}
#8 CACHED

#14 [internal] load build context
#14 sha256:e3a6a2a7307c9d8b79dc{}
#14 transferring context: 274.43kB 0.0s done
#14 DONE 0.0s

#9 [ 6/12] RUN pip install mlflow==1.19.0
#9 sha256:d1629c303a1736432{}
#9 0.665 /bin/sh: 1: pip: not found
#9 ERROR: executor failed running [/bin/sh -c pip install mlflow==1.19.0]: exit code: 127
------
 > [ 6/12] RUN pip install mlflow==1.19.0:
------
executor failed running [/bin/sh -c pip install mlflow==1.19.0]: exit code: 127
```
PS: I'm just redacting some sha's for readability!

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [x] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by harupy** [comment]
Hi @murilommen Thanks for reporting this issue. I'm trying to reproduce the issue.

```
% export DOCKER_BUILDKIT=1 
% mlflow models build-docker -m 'runs:/deea0f09c0054316b0cbc259bf313fc4/model' -n my-mlflow-image
...
#1 [internal] load .dockerignore
#1 transferring context: 2B done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 1.44kB done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/ubuntu:18.04
#3 DONE 0.0s

#4 [1/13] FROM docker.io/library/ubuntu:18.04
#15 DONE 0.0s

#5 [2/13] RUN apt-get -y update && apt-get install -y --no-install-recommen...
#5 DONE 130.5s

#6 [3/13] RUN curl -L https://repo.anaconda.com/miniconda/Miniconda3-latest...
#6 DONE 28.9s

#7 [4/13] RUN bash ./miniconda.sh -b -p /miniconda; rm ./miniconda.sh;
#7 DONE 9.5s

#8 [5/13] WORKDIR /opt/mlflow
#8 DONE 0.0s

#9 [6/13] RUN which pip
#9 0.177 /miniconda/bin/pip

...
```

**Comment by harupy** [comment]
`pip: not found` indicates that `/miniconda/bin` is not added to `PATH`.

https://github.com/mlflow/mlflow/blob/b125aae1aea5c880e8b2c052222de4a35d8fa787/mlflow/models/docker_utils.py#L34

**Comment by murilommen** [comment]
Hi @harupy thanks for the insights! 
I wasn't explicitly declaring the Docker BuildKit, but perhaps it was already running with it 😄 

I've modified the `docker_builder.py` using `&&` instead of the semi-colon as your PR resolves to, but it still returned me an error when building anaconda (`bash ./miniconda.sh`), which led me to thinking this is an issue for later macbooks with the M1 chip (my case), since there is no anaconda for the ARM64 architecture yet (that I'm aware of).

Interestingly enough, when modifying the `docker_utils.py` to this:
```python
proc = Popen(
            ["docker", "build", "-t", image_name, "-f", "Dockerfile", "." , "--platform=linux/amd64"],
```
It has successfully installed the necessary dependencies! 

Thanks again for the help here, I can open a PR to enforce this behavior if you think it is a good idea.

**Comment by harupy** [comment]
Thanks for the investigation. I tried the miniconda installation on my M1 MacBook Pro and got the following error:

```
 > [4/4] RUN bash ./miniconda.sh -b -p /miniconda:                                                                                                                                   
#7 0.159 PREFIX=/miniconda                                                                                                                                                           
#7 0.403 Unpacking payload ...                                                                                                                                                       
#7 0.404 /lib64/ld-linux-x86-64.so.2: No such file or directory
#7 0.406 /lib64/ld-linux-x86-64.so.2: No such file or directory
```

Did you get the same error?

**Comment by harupy** [comment]
Confirmed this works:

```
FROM ubuntu:18.04
RUN apt-get -y update && apt-get install -y --no-install-recommends curl ca-certificates
RUN curl -L https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-aarch64.sh >> miniconda.sh

RUN bash ./miniconda.sh -b -p /miniconda
ENV PATH="/miniconda/bin:$PATH"
```

```
% docker build -t issue-4561 .
[+] Building 1.0s (8/8) FINISHED                                                                                                                                                     
 => [internal] load build definition from Dockerfile                                                                                                                            0.0s
 => => transferring dockerfile: 324B                                                                                                                                            0.0s
 => [internal] load .dockerignore                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                 0.0s
 => [internal] load metadata for docker.io/library/ubuntu:18.04                                                                                                                 0.9s
 => [1/4] FROM docker.io/library/ubuntu:18.04@sha256:3b8692dc4474d4f6043fae285676699361792ce1828e22b1b57367b5c05457e3                                                           0.0s
 => CACHED [2/4] RUN apt-get -y update && apt-get install -y --no-install-recommends curl ca-certificates                                                                       0.0s
 => CACHED [3/4] RUN curl -L https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-aarch64.sh >> miniconda.sh                                                             0.0s
 => CACHED [4/4] RUN bash ./miniconda.sh -b -p /miniconda                                                                                                                       0.0s
 => exporting to image                                                                                                                                                          0.0s
 => => exporting layers                                                                                                                                                         0.0s
 => => writing image sha256:befce800f11447794371eced6618eae6fc39f65db1ababb0c82cdaea6c7c8f6e                                                                                    0.0s
 => => naming to docker.io/library/issue-4561                                                                                                                                   0.0s
                                                                                                                                                                                     
% docker run issue-4561 pip

Usage:   
  pip <command> [options]
```

**Comment by murilommen** [comment]
@harupy precisely! and that is where I had to change the `Popen` bit to include the `--platform=linux/amd64`

**Comment by murilommen** [comment]
maybe too obvious, but you will have to clear your builder cache in case you retry this

**Comment by harupy** [comment]
Tried the build with `--no-cache`:

```
% docker build -t issue-4561 --no-cache  .
[+] Building 35.8s (9/9) FINISHED                                                                                                                                                    
 => [internal] load build definition from Dockerfile                                                                                                                            0.0s
 => => transferring dockerfile: 324B                                                                                                                                            0.0s
 => [internal] load .dockerignore                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                 0.0s
 => [internal] load metadata for docker.io/library/ubuntu:18.04                                                                                                                 2.1s
 => [auth] library/ubuntu:pull token for registry-1.docker.io                                                                                                                   0.0s
 => CACHED [1/4] FROM docker.io/library/ubuntu:18.04@sha256:3b8692dc4474d4f6043fae285676699361792ce1828e22b1b57367b5c05457e3                                                    0.0s
 => [2/4] RUN apt-get -y update && apt-get install -y --no-install-recommends curl ca-certificates                                                                             11.1s
 => [3/4] RUN curl -L https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-aarch64.sh >> miniconda.sh                                                                   10.5s
 => [4/4] RUN bash ./miniconda.sh -b -p /miniconda                                                                                                                             10.2s 
 => exporting to image                                                                                                                                                          1.7s 
 => => exporting layers                                                                                                                                                         1.7s 
 => => writing image sha256:fead531d9add352f4c47574b3bb12c2ed1f7e309b70a428ca39d7adee4a9420f                                                                                    0.0s 
 => => naming to docker.io/library/issue-4561   
```

**Comment by harupy** [comment]
@murilommen Does `Miniconda3-latest-Linux-aarch64.sh` also work in your environmen?

**Comment by harupy** [comment]
Looks like we have two approaches to fix this issue:

1. Use `--platform=linux/amd64` option when building an image.
2. Use `Miniconda3-latest-Linux-aarch64.sh` if `arch` returns `aarch64`.

**Comment by murilommen** [comment]
@harupy Gonna try this later on today an i'll let you know. Thanks again!

**Comment by murilommen** [comment]
@harupy sorry for the long delay, but I just wanted to know: should we close this or open a PR to explicitly inform the platform/Miniconda version?

**Comment by dbczumar** [comment]
@harupy Any updates here?

**Comment by harupy** [comment]
@murilommen I think we should open a PR :)

**Comment by murilommen** [comment]
@harupy opening it right now! thanks

**Comment by murilommen** [comment]
@harupy saw that you fixed the last part on my behalf, thank you so much for your support! 😃 on my next contribution i will improve my commits management and try not to do it on a busy schedule again. 
should we close this issue?

**Comment by harupy** [comment]
@murilommen Yes, thanks for the contribution!

**Comment by harupy** [comment]
Fixed by #4688
